Security JPA implementation

FroMage opened the pull request:
This is for #4347, and it's Almost Done™ (Famous Last Words©).

You can define your security user entity like this:

```java
@UserDefinition
@Entity
public class PanacheUserEntity extends PanacheEntity {
    @Username
    public String name;

    @Password
    public String pass;

    @Roles
    public String roles;
}
```

Or with multiple tables for roles:

```java
@UserDefinition
@Entity
public class PanacheUserEntity extends PanacheEntity {
    @Username
    public String name;

    @Password
    public String pass;

    @ManyToMany
    @Roles
    public List<PanacheRoleEntity> roles = new ArrayList<>();
}

@Entity
public class PanacheRoleEntity extends PanacheEntity {

    @ManyToMany(mappedBy = "roles")
    public List<PanacheUserEntity> users;

    @RolesValue
    public String role;

}
```

- Supported role types of types `String`, `Collection/List/Set<String>`, `Collection/List/Set<role-entity>`, which can all contain comma-separated strings of roles.
- Supports plain Hibernate ORM or ORM with Panache
- No config required
- Supports `@NaturalId` on the same field as `@Username`
- Works _before_ the request, using `JPAConfig`'s support to get `EntityManagerFactory`
- Supports clear and MCF passwords

Under the hood, it generates this sort of `IdentityProvider` implementation:

```java
@Singleton
public class UserEntityIdentityProvider implements JpaIdentityProvider {

    @Override
    public SecurityIdentity authenticate(EntityManager em,
            UsernamePasswordAuthenticationRequest request) {

        // With Natural Id
        Session session = em.unwrap(Session.class);
        SimpleNaturalIdLoadAccess<PlainUserEntity> naturalIdLoadAccess = session.bySimpleNaturalId(PlainUserEntity.class);
        PlainUserEntity user = naturalIdLoadAccess.load(request.getUsername());
        // With normal query
        //        Query query = em.createQuery("FROM PlainUserEntity WHERE name = :name");
        //        query.setParameter("name", request.getUsername());
        //        PlainUserEntity user = getSingleUser(query);
        if (user == null)
            return null;

        // for MCF:
        //               Password storedPassword = getMcfPasword(user.pass);
        // for clear:
        Password storedPassword = getClearPassword(user.pass);

        QuarkusSecurityIdentity.Builder builder = checkPassword(storedPassword, request);

        addRoles(builder, user.role);
        return builder.build();
    }

}
```

So overall it's pretty complete, but I have a few questions before I'm ready to merge (besides the lack of docs):

- I default to MCF (#5667) passwords in single column (which supports bcrypt and many others), with clear text password being supported. I think this is fine for the start, right?
- I wanted to only require a single `@UserDefinition` annotation and default to looking up fields if they're named "user", "username", "pass", "password", "passwd", etc… but then I wondered if that would not be a security risk if a user adds a `@UserDefinition` annotation to his class and we pick up `user` for the username, but it turns out that in his model, this is not a primary key column and we should actually use the `email` column, so this could be a dangerous risk so I didn't do it. It feels lame to have to use 4 annotations, though, so WDYT? I'd love to use convention over configuration here.

FroMage committed:
Security JPA implementation

Sanne commented:
Some questions / ideas

- do you enforce/expect/require uniqueness of userids in any way?

- Should you load credentials by natural id? The great benefit is you'd not have to generate an HQL query - and parse them each time it's used. I think we should encourage people to use naturalids, or "process" their entities to enable the feature.

- I think I'd allow people to use a named query rather than hard code the query strategy. I guess this could be a future enhancement, but you might want to need take it into account so to allow evolution of the configuration-less use case?

- if you want this enabled and be configuration-less: having just the model won't be enough, you will need at least the datasource configuration.

- Do you have a plan for how this will look like when multiple persistence units / multiple datasources will be allowed?

- I'm confused that you load the credentials model just by userid and then seem to delegate checking the password to other business logic; I remember most other systems regularly require for both the userid and the password to match what's in the store before even allowing to load this. Essentially that's why most systems can't tell if the user exists vs the user exists but you don't have the right password - I don't know why that's the standard but I wonder if that's an important guideline?

- what's the expected semantics of `io.quarkus.elytron.security.jpa.UserEntityCredentialObtainer#obtain` ? It's returning managed entities, it would be far more efficient to use a stateless session. Clearly, not necessary for the implementation in integration testing but it would be good to apply this when generating an implementation. Speaking of, why have all that complexity expressed as gizmo code, couldn't you have a simple abstract class to extend via code generation? Might be easier to make such improvements in the future.

 - should "clear text" really be an option?

FroMage commented:
Thanks for the great feedback!

> do you enforce/expect/require uniqueness of userids in any way?

I don't, but I can't imagine a user lookup string being anything else than unique. I would rather avoid inspecting JPA annotations as much as possible because I'm sure I'll get them wrong (field/getter/setter/inheritance). What I can do is when generating the query, request at most 2 results and throw if `> 1`?  WDYT? Or do you advise I check for a unicity constraint? How?

> Should you load credentials by natural id? The great benefit is you'd not have to generate an HQL query - and parse them each time it's used. I think we should encourage people to use naturalids, or "process" their entities to enable the feature.

I can totally do that, good idea. Given an annotated field or getter, can I rely on the presence of `@NaturalId` on it? Do I need to check if that annotation is present on any other field to avoid composite natural ids?

> I think I'd allow people to use a named query rather than hard code the query strategy. I guess this could be a future enhancement, but you might want to need take it into account so to allow evolution of the configuration-less use case?

Named query for the entity loading, I guess, right? Not to return all the separate fields (we have the security-jdbc extension for something like that). I guess I can add a `@UserDefinition(namedQuery = "foo")` parameter which would make `@Username` optional. I assume that named query has a single username parameter, right?

> if you want this enabled and be configuration-less: having just the model won't be enough, you will need at least the datasource configuration.

Yes, but it was more a question of wether we need `security.jpa.enabled=true` or not. That's the only security config required ATM, and I think it should not be required.

> Do you have a plan for how this will look like when multiple persistence units / multiple datasources will be allowed?

Not more than for Hibernate ORM with Panache: I expect that alternate PU/DS will be specifiable as a package or entity annotation and we'll be able to use that.

> I'm confused that you load the credentials model just by userid and then seem to delegate checking the password to other business logic; I remember most other systems regularly require for both the userid and the password to match what's in the store before even allowing to load this. Essentially that's why most systems can't tell if the user exists vs the user exists but you don't have the right password - I don't know why that's the standard but I wonder if that's an important guideline?

So, I did wonder about that too, because I always looked users up by both user/pass, but Elytron does not. This is how the JDBC implementation works as well and the underlying API seems to force this. I just followed.

what's the expected semantics of io.quarkus.elytron.security.jpa.UserEntityCredentialObtainer#obtain ? It's returning managed entities, it would be far more efficient to use a stateless session. Clearly, not necessary for the implementation in integration testing but it would be good to apply this when generating an implementation. Speaking of, why have all that complexity expressed as gizmo code, couldn't you have a simple abstract class to extend via code generation? Might be easier to make such improvements in the future.

> No, this method fills a POJO with the results of the JPA session/query. It could use a stateless session if I knew how to use it. Would it be more efficient than this (this is what we generate):

```java
    @Override
    public CredentialHolder obtain(EntityManager entityManager, String username) {
        Query query = entityManager.createQuery("FROM PlainUserEntity WHERE name = :name");
        query.setParameter("name", username);
        List<PlainUserEntity> results = (List<PlainUserEntity>) query.getResultList();
        if (results.isEmpty())
            return null;
        // FIXME: more than one result?
        PlainUserEntity user = results.get(0);

        CredentialHolder ret = new CredentialHolder();

        ret.username = username;

        ret.setClearPassword(user.pass);

        ret.addRoles(user.role);
        return ret;
    }
```

> should "clear text" really be an option?

Yes, for demos. The other security impls support it too.

> setClearPassword .. is this implying it's a requirement to use symmetric cryptography?

Well, it's a clear text password, that's all you can infer. But clear text is indeed symmetric.

stuartwdouglas commented:
I have a few comments around this:

- What does Elytron add here? The existing Elytron extensions are here as they allowed us to re-use the existing Elytron realms while still integrating with our common security architecture. If you are providing a new Elytron realm then IMHO this should just be a new IdentityProvider rather than an Elytron realm.

- How does a user add an entity here in terms of setting the password? One approach would be to automagically transform the entity so that if you set the password to a plaintext string then we automatically hash it before flushing, but I think that is probably a bit too magic. A simpler approach would be to provide a PasswordService bean that allows the user to turn a plaintext password into a DB password. 

- I don't really like CredentialObtainer, IMHO it is to close to IdentityProvider. e.g. the example above is broadly similar to the following code:

```java
public class TestIdentityProvider implements IdentityProvider<UsernamePasswordAuthenticationRequest> {

    @Inject
    EntityManagerFactory emf;
    
    @Inject
    PasswordService passwordService;

    @Override
    public Class<UsernamePasswordAuthenticationRequest> getRequestType() {
        return UsernamePasswordAuthenticationRequest.class;
    }

    @Override
    public CompletionStage<SecurityIdentity> authenticate(UsernamePasswordAuthenticationRequest request,
            AuthenticationRequestContext context) {
        return context.runBlocking(new Supplier<SecurityIdentity>() {
            @Override
            public SecurityIdentity get() {
                EntityManager entityManager = emf.createEntityManager();
                try {
                    Query query = entityManager.createQuery("FROM PlainUserEntity WHERE name = :name");
                    query.setParameter("name", request.getUsername());
                    List<PlainUserEntity> results = (List<PlainUserEntity>) query.getResultList();
                    if (results.isEmpty())
                        return null;
                    // FIXME: more than one result?
                    PlainUserEntity user = results.get(0);
                    passwordService.verify(user.pass, request.getPassword());
                    return QuarkusSecurityIdentity.builder()
                            .setPrincipal(new QuarkusPrincipal(request.getUsername()))
                            .addRole(user.role)
                            .addCredential(request.getPassword())
                            .build();
                } finally {
                    entityManager.close();
                }

            }
        });
    }
}
```

I think this mostly just adds extra abstraction around our core API's.

Sanne commented:
> Thanks for the great feedback!

thanks :)

> > do you enforce/expect/require uniqueness of userids in any way?
> 
> I don't, but I can't imagine a user lookup string being anything else than unique. I would rather avoid inspecting JPA annotations as much as possible because I'm sure I'll get them wrong (field/getter/setter/inheritance). What I can do is when generating the query, request at most 2 results and throw if `> 1`? WDYT? Or do you advise I check for a unicity constraint? How?

If you use the natural id it enforces uniqueness. You'd still need to enforce the user to actually make use of the `@NaturalId` annotation but I think that would come ... natural. You wouldn't have to inspect them yourself, other than possibly check there's one to provide a great error message, but even if you don't, if there isn't one it would fail when attempting to load the entity by natural id.

> > Should you load credentials by natural id? The great benefit is you'd not have to generate an HQL query - and parse them each time it's used. I think we should encourage people to use naturalids, or "process" their entities to enable the feature.
> 
> I can totally do that, good idea. Given an annotated field or getter, can I rely on the presence of `@NaturalId` on it? Do I need to check if that annotation is present on any other field to avoid composite natural ids?

Hadn't thought about composite natural ids. Isn't it a given that you need the "unique id" to be the username?
I'd say that if users need to map a complex existing schema doing fancy stuffl like encoding the username into multiple columns, enforce them to still map a single `@NaturalId` property, they can always declare it as a composite embedded object and workaround such limitations.

That way your API also needs to only deal with this single object as id, rather than have alternative code to deal with tuplets.

> 
> > I think I'd allow people to use a named query rather than hard code the query strategy. I guess this could be a future enhancement, but you might want to need take it into account so to allow evolution of the configuration-less use case?
> 
> Named query for the entity loading, I guess, right? Not to return all the separate fields (we have the security-jdbc extension for something like that). I guess I can add a `@UserDefinition(namedQuery = "foo")` parameter which would make `@Username` optional. I assume that named query has a single username parameter, right?

A NamedQuery is very flexible, it can take as many parameters as the user wants.

But I just mentioned them as alternative to natural-ids. I'd say choose one approach, no need to go overboard with flexibility.

> 
> > if you want this enabled and be configuration-less: having just the model won't be enough, you will need at least the datasource configuration.
> 
> Yes, but it was more a question of wether we need `security.jpa.enabled=true` or not. That's the only security config required ATM, and I think it should not be required.

I don't think it can be the default. For example, even if there was a default datasource to bind to, it might not be mean for you to create additional tables in there without explicit opt-in, especially as they might end up storing sensitive information.

> > Do you have a plan for how this will look like when multiple persistence units / multiple datasources will be allowed?
> 
> Not more than for Hibernate ORM with Panache: I expect that alternate PU/DS will be specifiable as a package or entity annotation and we'll be able to use that.

It won't be an entity annotation, that would not be user friendly at all. I hope to write down a proposal soon and discuss about it.. just please keep in mind it will need to play nicely with it.

> 
> > I'm confused that you load the credentials model just by userid and then seem to delegate checking the password to other business logic; I remember most other systems regularly require for both the userid and the password to match what's in the store before even allowing to load this. Essentially that's why most systems can't tell if the user exists vs the user exists but you don't have the right password - I don't know why that's the standard but I wonder if that's an important guideline?
> 
> So, I did wonder about that too, because I always looked users up by both user/pass, but Elytron does not. This is how the JDBC implementation works as well and the underlying API seems to force this. I just followed.

Ok, was just wondering but it's not my field ./

> 
> what's the expected semantics of io.quarkus.elytron.security.jpa.UserEntityCredentialObtainer#obtain ? It's returning managed entities, it would be far more efficient to use a stateless session. Clearly, not necessary for the implementation in integration testing but it would be good to apply this when generating an implementation.
> 
> No, this method fills a POJO with the results of the JPA session/query. It could use a stateless session if I knew how to use it. Would it be more efficient than this (this is what we generate):

Yes it would. A lot.

> 
> ```java
>     @Override
>     public CredentialHolder obtain(EntityManager entityManager, String username) {
>         Query query = entityManager.createQuery("FROM PlainUserEntity WHERE name = :name");
>         query.setParameter("name", username);
>         List<PlainUserEntity> results = (List<PlainUserEntity>) query.getResultList();
>         if (results.isEmpty())
>             return null;
>         // FIXME: more than one result?
>         PlainUserEntity user = results.get(0);
> 
>         CredentialHolder ret = new CredentialHolder();
> 
>         ret.username = username;
> 
>         ret.setClearPassword(user.pass);
> 
>         ret.addRoles(user.role);
>         return ret;
>     }
> ```
> 
> > Speaking of, why have all that complexity expressed as gizmo code, couldn't you have a simple abstract class to extend via code generation? Might be easier to make such improvements in the future.
> 
> I do, this code above is what I generate. It mostly deals with reading fields/getters into the POJO. I could abstract the query part, but it seemed so trival that I didn't think it worth abstracting more.
> 
> > should "clear text" really be an option?
> 
> Yes, for demos. The other security impls support it too.

Ok
 
> > setClearPassword .. is this implying it's a requirement to use symmetric cryptography?
> 
> Well, it's a clear text password, that's all you can infer. But clear text is indeed symmetric.

ok so this method is specific to the "clear text" strategy? Just wondering if it wouldn't get in the way in case of non-symmetric strategies potentially coming later.

Sanne commented:
P.S. I agree with Stuart that `CredentialObtainer` seems weird looking :)

FroMage commented:
> What does Elytron add here? The existing Elytron extensions are here as they allowed us to re-use the existing Elytron realms while still integrating with our common security architecture. If you are providing a new Elytron realm then IMHO this should just be a new IdentityProvider rather than an Elytron realm.

I started as a copy of elytron-jpa, then trimmed from there. It still depends on elytron for the specifying the credentials. Can I really do an implementation that doesn't depend on it? Note that I'm not familiar at all with `IdentityProvider`. Where is it documented?

> How does a user add an entity here in terms of setting the password? One approach would be to automagically transform the entity so that if you set the password to a plaintext string then we automatically hash it before flushing, but I think that is probably a bit too magic. A simpler approach would be to provide a PasswordService bean that allows the user to turn a plaintext password into a DB password.

That's not our business. We're in the business of reading entities, not writing them. The user is responsible for storing the entities, and they can use the elytron API to create the right hashes. That's pretty trivial, and I really don't think we should meddle with entity storage logic.

FroMage commented:
> I don't really like CredentialObtainer, IMHO it is to close to IdentityProvider. e.g. the example above is broadly similar to the following code:

```java
public class TestIdentityProvider implements IdentityProvider<UsernamePasswordAuthenticationRequest> {

    @Inject
    EntityManagerFactory emf;
    
    @Inject
    PasswordService passwordService;

    @Override
    public Class<UsernamePasswordAuthenticationRequest> getRequestType() {
        return UsernamePasswordAuthenticationRequest.class;
    }

    @Override
    public CompletionStage<SecurityIdentity> authenticate(UsernamePasswordAuthenticationRequest request,
            AuthenticationRequestContext context) {
        return context.runBlocking(new Supplier<SecurityIdentity>() {
            @Override
            public SecurityIdentity get() {
                EntityManager entityManager = emf.createEntityManager();
                try {
                    Query query = entityManager.createQuery("FROM PlainUserEntity WHERE name = :name");
                    query.setParameter("name", request.getUsername());
                    List<PlainUserEntity> results = (List<PlainUserEntity>) query.getResultList();
                    if (results.isEmpty())
                        return null;
                    // FIXME: more than one result?
                    PlainUserEntity user = results.get(0);
                    passwordService.verify(user.pass, request.getPassword());
                    return QuarkusSecurityIdentity.builder()
                            .setPrincipal(new QuarkusPrincipal(request.getUsername()))
                            .addRole(user.role)
                            .addCredential(request.getPassword())
                            .build();
                } finally {
                    entityManager.close();
                }

            }
        });
    }
}
```

OK, I'll admit I'd never heard of this API to define a pluggable security provider. 
- Is it documented somewhere? 
- Is is auto-discovered (one of the nice points of `CredentialsObtainer`)?
- What's `PasswordService`?
- How do we specify anything else than plan-text passwords?

Otherwise, I agree it's the same idea as `CredentialsObtainer`, except way more verbose, but I can get around that by providing a supertype to override. The one thing I don't understand is whether it does the password checking or not. `CredentialsObtainer` does not: it's only responsible for data extraction.

FroMage commented:
@Sanne 

> Hadn't thought about composite natural ids. Isn't it a given that you need the "unique id" to be the username?
> I'd say that if users need to map a complex existing schema doing fancy stuffl like encoding the username into multiple columns, enforce them to still map a single @NaturalId property, they can always declare it as a composite embedded object and workaround such limitations.

> That way your API also needs to only deal with this single object as id, rather than have alternative code to deal with tuplets.

We really look entities up by username, which can only be a `String`, so I don't know how composite embedded objects can work.

> A NamedQuery is very flexible, it can take as many parameters as the user wants.

OK, but the only dynamic param we can pass is the username. Unless we have a source of other params, but I wonder which.

> But I just mentioned them as alternative to natural-ids. I'd say choose one approach, no need to go overboard with flexibility.

OK let's skip it for now then ;)

> I don't think it can be the default. For example, even if there was a default datasource to bind to, it might not be meant for you to create additional tables in there without explicit opt-in, especially as they might end up storing sensitive information.

We're not creating any tables in the security extension. We're just reading how to map the user model to our security source.

>> No, this method fills a POJO with the results of the JPA session/query. It could use a stateless session if I knew how to use it. Would it be more efficient than this (this is what we generate):

> Yes it would. A lot.

Please show me equivalent code with stateless session then :)

>>> setClearPassword .. is this implying it's a requirement to use symmetric cryptography?

>> Well, it's a clear text password, that's all you can infer. But clear text is indeed symmetric.

> ok so this method is specific to the "clear text" strategy? Just wondering if it wouldn't get in the way in case of non-symmetric strategies potentially coming later.

The hashed password code we generate is:

```java
ret.setMdfPassword(user.pass);
```

The generated code depends on how the `@Password` annotation is declared.

Sanne commented:
ok so, summarizing as I feel you only have 2 open points still:

embeeded natural ids:
yes I also was thinking only for String, but you brought the composite-key possibility up. Ignore it then and assume there's a unique string?
That was my suggestion, as the user can bypass the limitation by using a more complex mapping I think - but also it feels very far fetched.

Just use NaturalId and expect a String for username.

> Please show me equivalent code with stateless session then :)

    try (StatelessSession statelessSession = sessionFactory.openStatelessSession()) {
        Person p = statelessSession.get( Person.class, id );
        return p;
    }

Unfortunatley I just noticed the stateless session API doesn't do natural id loading. Seems like an oversight, but no big deal just forget about my suggestion to use it and go for named queries instead?

Would be cool to add some helpers in Panache?

FroMage commented:
> Unfortunatley I just noticed the stateless session API doesn't do natural id loading. Seems like an oversight, but no big deal just forget about my suggestion to use it and go for named queries instead?

- Do stateless sessions allow me to have the roles in a relation?
- I can revert to `EntityManager` if using a `@NaturalId`
- I'd rather also keep the current option to just annotate a field with `@Username` for lookup: I find named queries are harder to write than just annotating the password field. I'd rather add the named query support later.

> Would be cool to add some helpers in Panache?

I think there's an open issue for this.

stuartwdouglas commented:
IdentityProvider is the core of our security API. It was documented in the security design doc from Paris (you were there when we designed it). 

Originally the plan was not to use Elytron, but due to time constraints the existing Elytron stuff was modified to just use our new core API's. This is an extensible API, the user can simply include it in their application and make it a CDI bean and it will be picked up.

This has no support for passwords though, so we should use the Elytron stuff for that, but ideally behind some kind of easy to use API that the users can just use directly. Elytron is modular, so you can include the password support without also including the realm support.

FroMage commented:
> IdentityProvider is the core of our security API. It was documented in the security design doc from Paris (you were there when we designed it).

I didn't mean to imply it was not justified. It's just that I forgot about this particular interface, and didn't read anything about it when I reviewed the security guides a few weeks ago. We should document it in our guides.

OK lemme rewrite that in terms of `IdentityProvider` then.

Sanne commented:
> > Unfortunatley I just noticed the stateless session API doesn't do natural id loading. Seems like an oversight, but no big deal just forget about my suggestion to use it and go for named queries instead?
> 
> * Do stateless sessions allow me to have the roles in a relation?
> * I can revert to `EntityManager` if using a `@NaturalId`
> * I'd rather also keep the current option to just annotate a field with `@Username` for lookup: I find named queries are harder to write than just annotating the password field. I'd rather add the named query support later.
> 
> > Would be cool to add some helpers in Panache?
> 
> I think there's an open issue for this.

Ok to go with NaturalId then. But no you won't be ableto use `EntityManager` as it doesn't have the method
https://docs.jboss.org/hibernate/orm/5.4/userguide/html_single/Hibernate_User_Guide.html#naturalid-api

You can get most of the performance benefits of a StatelessSession by setting flush mode to manual, and set the session in read-only mode.
I'd recommend doing it also for safety (not just performance) so to make sure you don't unintentionally modify the database.

FroMage commented:
OK I'll try.

nimo23 commented:
Why not using `Java EE 8 Security API` behind the scenes? It already has Built-in Identity Stores, e.g. `@DatabaseIdentityStoreDefinition`. Why not providing the `@Username`, `@Password`, `@Roles` an an extension of `Java EE 8 Security API`?

FroMage commented:
@Sanne I can't actually use a stateless session if my roles are in another table:

```
2019-12-04 17:19:51,217 ERROR [io.qua.ver.htt.run.QuarkusErrorHandler] (executor-thread-1) HTTP Request to /jaxrs-secured/parameterized-paths/my/banking/view failed, error id: 9bb728e4-477a-4961-9074-5ed2ff311af1-24: org.hibernate.LazyInitializationException: Unable to perform requested lazy initialization [io.quarkus.security.jpa.PanacheUserEntity.roles] - no session and settings disallow loading outside the Session
	at org.hibernate.bytecode.enhance.spi.interceptor.EnhancementHelper.throwLazyInitializationException(EnhancementHelper.java:196)
	at org.hibernate.bytecode.enhance.spi.interceptor.EnhancementHelper.performWork(EnhancementHelper.java:86)
	at org.hibernate.bytecode.enhance.spi.interceptor.LazyAttributeLoadingInterceptor.loadAttribute(LazyAttributeLoadingInterceptor.java:76)
	at org.hibernate.bytecode.enhance.spi.interceptor.LazyAttributeLoadingInterceptor.fetchAttribute(LazyAttributeLoadingInterceptor.java:72)
	at org.hibernate.bytecode.enhance.spi.interceptor.LazyAttributeLoadingInterceptor.handleRead(LazyAttributeLoadingInterceptor.java:53)
	at org.hibernate.bytecode.enhance.spi.interceptor.AbstractInterceptor.readObject(AbstractInterceptor.java:153)
	at io.quarkus.security.jpa.PanacheUserEntity.$$_hibernate_read_roles(PanacheUserEntity.java)
	at io.quarkus.security.jpa.PanacheUserEntity.getRoles(PanacheUserEntity.java)
	at io.quarkus.security.jpa.PanacheUserEntity__JpaIdentityProviderImpl.authenticate(PanacheUserEntity__JpaIdentityProviderImpl.zig:76)
```

Sanne commented:
@FroMage sure, see my last comment  - I thought we had agreed to not use stateless sessions but rather prefer using natural Ids ?

The need to handle relations is one of the reasons.

FroMage commented:
I guess I'm trying things out in comment order :(

Sanne commented:
very efficient :dancer: have fun :)

sorry about that, I hadn't initially thought about the roles being mapped there as well.

FroMage commented:
OK, I've added support for `@NaturalId` @Sanne (if it's not there it will revert to a normal query), and turned the session read-only with no flush. We'll deal with named queries if/when someone requests it.

@stuartwdouglas I've rewritten in terms of `IdentityProvider`.

As a result, I've stopped depending on `quarkus-security-elytron-deployment` (but still on the runtime because it contains SVM substs I need), and also there's no config anymore: if there's an annotated entity, it generates an `IdentityProvider`, if not: it doesn't do anything.

If you guys think it's fine, I'll write the docs.

FroMage commented:
@nimo23 
> It already has Built-in Identity Stores, e.g. @DatabaseIdentityStoreDefinition

Given this example:

```java
@DatabaseIdentityStoreDefinition(
    callerQuery = "#{'select password from caller where name = ?'}",
    groupsQuery = "select group_name from caller_groups where caller_name = ?",
    hashAlgorithm = Pbkdf2PasswordHash.class,
    priorityExpression = "#{100}",
    hashAlgorithmParameters = {
        "Pbkdf2PasswordHash.Iterations=3072",
        "${applicationConfig.dyna}"
    }
)
```

It looks like we already support this in our `quarkus-security-elytron-jdbc` extension. I don't think an annotation is the right place for this sort of config.

We could probably support our annotated entities on top of that, but I'd much rather plug into the existing Quarkus security APIs than EE security which comes with a lot of baggage.

FroMage commented:
It would be nice to get this in 1.1 @gsmet if I can get it reviewed and finished in time.

FroMage commented:
Just updated the PR description with latest features.

FroMage commented:
Rebased on master, let's see if it helps wrt CI. Without feedback from @Sanne and @stuartwdouglas I don't think it can make 1.1.

stuartwdouglas commented on extensions/security-jpa/deployment/src/main/java/io/quarkus/security/jpa/deployment/QuarkusSecurityJpaProcessor.java:
```diff
@@ -0,0 +1,370 @@
+
+            @Override
+            public Writer getSourceWriter(String className) {
+                return new OutputStreamWriter(System.err) {
```
Is this supposed to be here or was this a debugging thing? If it is not supposed to be there you can use GeneratedBeanGizmoAdaptor

FroMage replied:
I will remove this.

FroMage replied:
Removed

stuartwdouglas commented on extensions/security-jpa/runtime/src/main/java/io/quarkus/security/jpa/JpaRecorder.java:
```diff
@@ -0,0 +1,16 @@
+public class JpaRecorder {
+
+    public void registerSecurityProvider() {
+        Security.addProvider(new WildFlyElytronPasswordProvider());
```
This is also registered by io.quarkus.elytron.security.runtime.ElytronRecorder#registerPasswordProvider, I don't think we want it twice (not sure if it matters)? If registering it twice causes problems we probably need some build items to manage the provider registration.

FroMage replied:
IIRC it is conditionally added only if you have a realm, which is not the case here.

FroMage replied:
So it's not conditional anymore, but I'd have to make my extension depend on `quarkus-security-elytron` to get it. Do you think that's fine?

FroMage replied:
I removed my recorder and depend on that extension. I guess it should be fine.

dmlloyd replied:
We might consider switching this to be ServiceLoader-based.  All of the Elytron providers are available via `ServiceLoader.load(Provider.class)` normally; we could have a startup step where we register all providers on the class path at boot.

The JDK has already starting moving to this behavior in fact, and moving away from requiring global provider registration in some cases [1].  So really, we probably should look at enhancing the Elytron APIs to use this approach as well.

stuartwdouglas commented on extensions/security-jpa/runtime/src/main/java/io/quarkus/security/jpa/JpaSecurityConfig.java:
```diff
@@ -0,0 +1,17 @@
+ * A configuration object for JPA-based security.
+ */
+@ConfigRoot(name = "security.jpa", phase = ConfigPhase.BUILD_AND_RUN_TIME_FIXED)
+public class JpaSecurityConfig {
```
What is the point of the empty config? I think this should at least have the unit name.

FroMage replied:
Yeah, I will remove the empty config. I wasn't sure if I was going to need config or not, but I can't think of anything.

FroMage replied:
Removed

stuartwdouglas reviewed (changes requested):
I think it looks good, the only think I think it is missing (other than docs) is an easy way to generate the passwords. If I am implementing the 'new user' screen in my app there should be a really simple way to set the password on the entity, including generating the salt etc, I should not have to know anything about MCF passwords.

FroMage commented:
Finally got the stack trace of the failing test. This actually doesn't seem to be related to native, but the test passes in the IDE, but not on the Maven CLI.

```
sun.misc.Unsafe.park(Native Method)
java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
java.util.concurrent.CompletableFuture$Signaller.block(CompletableFuture.java:1693)
java.util.concurrent.ForkJoinPool.managedBlock(ForkJoinPool.java:3323)
java.util.concurrent.CompletableFuture.waitingGet(CompletableFuture.java:1729)
java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1895)
io.quarkus.runner.RuntimeClassLoader.findClass(RuntimeClassLoader.java:214)
io.quarkus.runner.RuntimeClassLoader.loadClass(RuntimeClassLoader.java:179)
java.lang.ClassLoader.loadClass(ClassLoader.java:357)
java.lang.Class.forName0(Native Method)
java.lang.Class.forName(Class.java:348)
org.objectweb.asm.ClassWriter.getCommonSuperClass(ClassWriter.java:935)
org.objectweb.asm.SymbolTable.addMergedType(SymbolTable.java:1200)
org.objectweb.asm.Frame.merge(Frame.java:1299)
org.objectweb.asm.Frame.merge(Frame.java:1197)
org.objectweb.asm.MethodWriter.computeAllFrames(MethodWriter.java:1607)
org.objectweb.asm.MethodWriter.visitMaxs(MethodWriter.java:1543)
org.objectweb.asm.ClassReader.readCode(ClassReader.java:2431)
org.objectweb.asm.ClassReader.readMethod(ClassReader.java:1283)
org.objectweb.asm.ClassReader.accept(ClassReader.java:688)
org.objectweb.asm.ClassReader.accept(ClassReader.java:400)
io.quarkus.hibernate.orm.deployment.HibernateEntityEnhancer$HibernateEnhancingClassVisitor.visitEnd(HibernateEntityEnhancer.java:67)
org.objectweb.asm.ClassReader.accept(ClassReader.java:692)
org.objectweb.asm.ClassReader.accept(ClassReader.java:400)
io.quarkus.runner.RuntimeClassLoader.handleTransform(RuntimeClassLoader.java:430)
io.quarkus.runner.RuntimeClassLoader.findClass(RuntimeClassLoader.java:225)
io.quarkus.runner.RuntimeClassLoader.loadClass(RuntimeClassLoader.java:179)
java.lang.ClassLoader.loadClass(ClassLoader.java:357)
java.lang.Class.forName0(Native Method)
java.lang.Class.forName(Class.java:348)
io.quarkus.hibernate.orm.runtime.service.FlatClassLoaderService.classForName(FlatClassLoaderService.java:36)
org.hibernate.boot.model.source.internal.annotations.AnnotationMetadataSourceProcessorImpl.<init>(AnnotationMetadataSourceProcessorImpl.java:109)
org.hibernate.boot.model.process.spi.MetadataBuildingProcess$1.<init>(MetadataBuildingProcess.java:155)
org.hibernate.boot.model.process.spi.MetadataBuildingProcess.complete(MetadataBuildingProcess.java:149)
io.quarkus.hibernate.orm.runtime.boot.FastBootMetadataBuilder.build(FastBootMetadataBuilder.java:342)
io.quarkus.hibernate.orm.runtime.PersistenceUnitsHolder.createMetadata(PersistenceUnitsHolder.java:111)
io.quarkus.hibernate.orm.runtime.PersistenceUnitsHolder.constructMetadataAdvance(PersistenceUnitsHolder.java:84)
io.quarkus.hibernate.orm.runtime.PersistenceUnitsHolder.initializeJpa(PersistenceUnitsHolder.java:47)
io.quarkus.hibernate.orm.runtime.HibernateOrmRecorder$4.created(HibernateOrmRecorder.java:75)
io.quarkus.arc.runtime.ArcRecorder.initBeanContainer(ArcRecorder.java:101)
io.quarkus.deployment.steps.ArcProcessor$generateResources24.deploy_0(ArcProcessor$generateResources24.zig:442)
io.quarkus.deployment.steps.ArcProcessor$generateResources24.deploy(ArcProcessor$generateResources24.zig:36)
io.quarkus.runner.ApplicationImpl.<clinit>(ApplicationImpl.zig:378)
java.lang.Class.forName0(Native Method)
```

Apparently it's a deadlock in the class loading, in an area (find common supertype) we recently touched, right @stuartwdouglas ?

FroMage commented:
Rebased on master which has a fix for that.

FroMage commented:
>  think it is missing (other than docs) is an easy way to generate the passwords. If I am implementing the 'new user' screen in my app there should be a really simple way to set the password on the entity, including generating the salt etc, I should not have to know anything about MCF passwords.

Done:

```java
        String testPassword = "fubar";

// this is the hash to store in DB        
        String testPasswordHash = JpaSecurityUtil.bcryptHash(testPassword);
        
        PasswordGuessEvidence correctPasswordEvidence = new PasswordGuessEvidence(testPassword.toCharArray());
        PasswordGuessEvidence incorrectPasswordEvidence = new PasswordGuessEvidence("stef".toCharArray());
        PasswordCredential producedPasswordCredential = new PasswordCredential(ModularCrypt.decode(testPasswordHash));
        Assertions.assertTrue(producedPasswordCredential.verify(correctPasswordEvidence));
        Assertions.assertFalse(producedPasswordCredential.verify(incorrectPasswordEvidence));

```

FroMage commented:
Tomorrow I'll write the docs and let's get rid of this PR.

FroMage commented:
I guess the `JpaSecurityUtil` file could do with a review by @dmlloyd :)

FroMage commented:
OK it's ready: it has javadocs, the util class, a guide, a quickstart, tests, the works.

sberyozkin commented:
@FroMage It looks good. What is the migration path for `quarkus-elytron-security-jdbc` users with regard to the password protection algorithms ? I.e, should the two extensions, "-jdbc" and this new one, support a common way to protect the passwords or it is not possible (referring here to the MCF passwords) ? thanks

FroMage commented:
@sberyozkin I don't know if we should have a migration strategy. I do think "clear text" should not have been the default and that bcrypt in MCF is the only default that has made sense for the last few years. I was very frustrated that the JDBC security seemed to force me to split my hashed password in many columns, but perhaps I was missing an option.

I'm not too tempted to add support for split passwords in this new extension because I think it's a bad design in many cases and wanted to wait for users to request it (or not). Perhaps we'l have no choice, but I'd rather not start with it to not encourage people to add complexity to their tables.

sberyozkin commented:
@FroMage Sure that sounds reasonable. Might make sense to mention in the docs why the new protection mechanism offered by this new extension should be preferred. thanks

FroMage commented:
@stuartwdouglas I think this is waiting for your review :)

danielpetisme commented on extensions/security-jpa/runtime/src/main/java/io/quarkus/security/jpa/runtime/JpaSecurityUtil.java:
```diff
@@ -0,0 +1,90 @@
+/**
+ * Utility class used to produce bcrypt hashes using the Modular Crypt Format.
+ */
+public class JpaSecurityUtil {
```
Based on #5113 and #5667 would it make sense to move an rename the class to something les "JPA related"? Like `io.quarkus.security.CryptoUtil` or similar?

FroMage replied:
That's a good point, I'll try to move it.

FroMage replied:
Given that it depends on Elytron and `quarkus-security` does not depend on it, I can move it to `quarkus-elytron-security-common` which ATM only holds the GraalVM substs that I need anyway.

I don't think it's worth adding it to ``quarkus-elytron-security` itself since it comes with baggage.

WDYT @stuartwdouglas that place would work for you?

stuartwdouglas replied:
Works for me.

FroMage replied:
Done. Had to move some things that init elytron/password and register them as service at runtime there, but I think it's reasonable: https://github.com/quarkusio/quarkus/pull/5660/commits/d80322935c72c81a61eb305d0de2013f3894501d

stuartwdouglas reviewed (approved):
Other than my minor comment I think this is good to go.

FroMage commented on core/deployment/src/main/java/io/quarkus/deployment/Capabilities.java:
```diff
@@ -24,6 +24,7 @@
     public static final String SECURITY = "io.quarkus.security";
     public static final String SECURITY_ELYTRON_OAUTH2 = "io.quarkus.elytron.security.oauth2";
     public static final String SECURITY_ELYTRON_JDBC = "io.quarkus.elytron.security.jdbc";
+    public static final String SECURITY_ELYTRON_JPA = "io.quarkus.elytron.security.jpa";
```
Ah, right.

FroMage replied:
Done.

sberyozkin commented:
@FroMage Hi Stephane, can you resolver the conflicts please ? Stuart is good with this PR so we can just merge it

FroMage commented:
Done. Perhaps needs a final review of https://github.com/quarkusio/quarkus/pull/5660/commits/d80322935c72c81a61eb305d0de2013f3894501d by whoever is responsible for `quarkus-elytron-security` and `quarkus-elytron-security-common`?

sberyozkin commented:
@FroMage It looks OK. Can you please squash the commits as well

stuartwdouglas commented:
There is a failure that may be related. I have seen is before related to registering security providers at static init time rather than runtime, but I am not sure why these changes would have caused it.

FroMage commented:
It must be related because infinispan is the only consumer of `quarkus-elytron-security-common` which I touched. But that NPE sure is not helping find the cause.

stuartwdouglas commented:
To debug stuff like this load up the graal sources, and run the native build with -Dquarkus.native.debug-build-process=true then attach to the native image compilation with a breakpoint at the problem line. I am pretty sure it is because something is registering a security provider at static init, but not sure why these changes would cause it.

FroMage commented:
It works here after I rebased on master and used the latest graalvm 19.3.1, so repushed and let's see what CI says.

gsmet reviewed (changes requested):
This needs a squash before merging.

gsmet commented on extensions/security-jpa/runtime/src/main/resources/META-INF/quarkus-extension.json:
```diff
@@ -0,0 +1,9 @@
+        "security",
+        "jpa"
+    ],
+    "guide": "https://quarkus.io/guides/security-jpa"
```
This should be a YAML file now.

FroMage replied:
OK thanks, fixed. Damn this was a long-standing PR :(

FroMage commented:
Actually, does it make sense to make the extension status `stable` @gsmet ? Should it be preview or something else?

FroMage commented:
OK switched extension status to preview. We should be able to merge this without re-running CI, no?

stuartwdouglas commented:
This needs a rebase

FroMage commented:
OK, here's another rebase.

FroMage commented:
This never ends. Now the tests fail with:

```
2020-01-23 09:31:50,082 ERROR [io.qua.ver.htt.run.QuarkusErrorHandler] (executor-thread-1) HTTP Request to /jaxrs-secured/roles-class failed, error id: 4afd1ea0-b20b-40f2-9278-ad328ca2aae0-8: java.lang.ClassCastException: org.wildfly.security.password.impl.PasswordFactorySpiImpl cannot be cast to org.wildfly.security.password.PasswordFactorySpi
```

@stuartwdouglas could it be related to the CP changes?

stuartwdouglas commented:
Probably, if I don't get to it tonight I will have a look tomorrow.

FroMage commented:
Thanks.

stuartwdouglas commented:
It's something to do with the password providers being loaded with the wrong class loader. It may be that we need to make sure Elytron is always loaded in the system class loader, or there may be a simpler fix, I will look tomorrow.

dmlloyd commented:
I think it's likely the opposite problem: there is more than one copy of the API JAR on the class path.  If we're supporting distinct class loaders, it's critically important that we do not allow classes to be duplicated between them.

In JBoss Modules we solved this by unconditionally using child-first delegation for every class loader, but also deleting any package path from the child that is also present in any parent, so it behaves like parent-first but without the drawbacks inherent in that delegation model.

stuartwdouglas commented:
Classes have to be duplicated, it is the old way we can support transformers, and make sure that our class path is 'clean' and not linked against old versions of API's that just happen to be in the parent class loader (e.g. gradle, although this one can likely be worked around in other ways, so is not a hard requirement).  

Basically the main requirements that lead to duplicated classes are:
- JUnit loads classes eagerly, so classes present in the system class loader cannot be reliably transformed, this needs them to be duplicated into a new class loader
- Also any dependency may also need to be transformed. These suffer from the same eager loading issues, but also need to be loaded into a 'flatish' class path, where they have two way visibility with other dependencies. 
- Isolation between deployment and runtime to allow for deployment to have additional dependencies. This was Sanne's use case, but this also means that the test environment is much more similar to the production environment, as static state left over from deployment is not visible at runtime.

Some things obviously can't be duplicated though and need to be in the system class loader (e.g. JBoss Logging), so there is a facility to allow that. I think it will likely turn out that the solution here is to force security artifacts to be parent first.

stuartwdouglas commented:
@FroMage I have pushed a commit that should help

FroMage commented:
Well, now it's the Windows build that fails:

```
[ERROR] testThatTheApplicationIsReloadedOnKotlinChange  Time elapsed: 121.786 s  <<< ERROR!
```

That can't be related to this PR, can it?

FroMage commented:
YAR (Yet Another Rebase).

FroMage commented:
YAR.

gsmet commented on bom/deployment/pom.xml:
```diff
@@ -612,6 +612,11 @@
             </dependency>
+            <dependency>
+                <groupId>io.quarkus</groupId>
+                <artifactId>quarkus-security-jpa-deployment</artifactId>
```
I think the name here is incorrect. At least it's inconsistent with the other `elytron-security` extensions (they have elytron in the name).

stuartwdouglas replied:
So this is a bit of a weird one, this is not really an Elytron extension even though it does use some of the Elytron password functionality, it does not use the Elytron realm mechanism.

FroMage replied:
It's only using Elytron under the hood for bCrypt, which is an implementation detail.

gsmet replied:
@stuartwdouglas apparently you were the one asking for a rename. I'm just a bit concerned that all the other security modules have Elytron in the name and I'm not really sure they depend more on Elytron than this one but maybe I'm mistaken? Not being consistent is an issue IMHO. I wasn't really fond of having Elytron in the name to begin with but I'm not sure it's a good idea to not be consistent.

stuartwdouglas replied:
OIDC does not have Elytron in the name, as it does not use Elytron. Elytron is not the core of our security infrastructure, we have some modules that use it because we did not have the time or resources to implement some of that stuff ourselves, but Elytron is inherently blocking.

gsmet commented on bom/runtime/pom.xml:
```diff
@@ -733,6 +733,11 @@
             </dependency>
+            <dependency>
+                <groupId>io.quarkus</groupId>
+                <artifactId>quarkus-security-jpa</artifactId>
```
Same here.

gsmet commented on docs/src/main/asciidoc/security-jpa.adoc:
```diff
@@ -0,0 +1,360 @@
+mvn io.quarkus:quarkus-maven-plugin:{quarkus-version}:create \
+    -DprojectGroupId=org.acme \
+    -DprojectArtifactId=security-jpa-quickstart \
+    -Dextensions="elytron-security-jpa, jdbc-postgresql, resteasy, hibernate-orm-panache"
```
And see here you have the right name.

FroMage replied:
Agree they should be in sync, this one's a mistake.

gsmet replied:
You have another occurrence in the doc then IIRC

gsmet commented on docs/src/main/asciidoc/security-jpa.adoc:
```diff
@@ -0,0 +1,360 @@
+import io.quarkus.security.jpa.runtime.Password;
+import io.quarkus.security.jpa.runtime.Roles;
+import io.quarkus.security.jpa.runtime.UserDefinition;
+import io.quarkus.security.jpa.runtime.Username;
```
Everything that is user API should be at the root, not in the `runtime` package. `runtime` is an internal package.

The package should also be `elytron.security.jpa` to be consistent with the other Elytron extensions.

FroMage replied:
Oh, I didn't know that. Will move.

gsmet commented on docs/src/main/asciidoc/security-jpa.adoc:
```diff
@@ -0,0 +1,360 @@
+
+<1> This annotation must be present on a single entity. It can be a regular Hibernate ORM entity or a Hibernate ORM with Panache entity as in this example.
+<2> This indicates the field used for the user name.
+<3> This indicates the field used for the password. This defaults to using bcrypt hashed passwords, but you can add also configure it for clear text passwords.
```
```suggestion
<3> This indicates the field used for the password. This defaults to using bcrypt hashed passwords, but you can also configure it for clear text passwords.
```

gsmet commented on docs/src/main/asciidoc/security-jpa.adoc:
```diff
@@ -0,0 +1,360 @@
+quarkus.hibernate-orm.database.generation=drop-and-create
+----
+
+In our context, we are using PostgreSQL as identity store. The database schema is created by hibernate automatically
```
```suggestion
In our context, we are using PostgreSQL as identity store. The database schema is created by Hibernate ORM automatically
```

gsmet commented on docs/src/main/asciidoc/security-jpa.adoc:
```diff
@@ -0,0 +1,360 @@
+
+[NOTE]
+====
+It is probably useless but we kindly remind you that you must not store clear-text passwords in production environment ;-).
```
```suggestion
It is probably useless but we kindly remind you that you must not store clear-text passwords in production environments ;-).
```

gsmet commented on docs/src/main/asciidoc/security-jpa.adoc:
```diff
@@ -0,0 +1,360 @@
+By default, we consider passwords to be stored hashed with https://en.wikipedia.org/wiki/Bcrypt[bcrypt] under the 
+https://en.wikipedia.org/wiki/Crypt_(C)[Modular Crypt Format] (MCF).
+
+When you need to create such a hashed password we provide the convenient `String BcryptUtil.bcryptHash(String password)`
```
If we expect the user to use it, we will need that one to be API too so outside of the runtime package.

FroMage replied:
OK

gsmet commented on docs/src/main/asciidoc/security.adoc:
```diff
@@ -22,6 +22,9 @@ in order for Quarkus to know how to find the authentication information to check
 |link:security-properties[quarkus-elytron-security-properties-file]
 |Provides support for simple properties files that can be used for testing security. This supports both embedding user info in `application.properties` and standalone properties files.
 
+|link:security-jpa[quarkus-security-jpa]
```
```suggestion
|link:security-jpa[quarkus-elytron-security-jpa]
```

gsmet commented on extensions/elytron-security-common/deployment/pom.xml:
```diff
@@ -18,6 +18,10 @@
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-elytron-security-common</artifactId>
+        </dependency>
```
Good catch.

gsmet commented on extensions/elytron-security-common/runtime/src/main/java/io/quarkus/elytron/security/common/runtime/BcryptUtil.java:
```diff
@@ -0,0 +1,90 @@
+
+        PasswordFactory passwordFactory;
+        try {
+            passwordFactory = PasswordFactory.getInstance(BCryptPassword.ALGORITHM_BCRYPT, provider);
```
Can't that be static and called once and for all? It might or might not be slow.

FroMage replied:
@dmlloyd I'm not sure. Can it? It probably will depend on the security provider being registered in the static runtime init before this one, but I'm not sure we can easily guarantee order, right?

gsmet commented on extensions/elytron-security-common/runtime/src/main/java/io/quarkus/elytron/security/common/runtime/BcryptUtil.java:
```diff
@@ -0,0 +1,90 @@
+        Objects.requireNonNull(password, "password is required");
+        Objects.requireNonNull(salt, "salt is required");
+        if (salt.length != BCryptPassword.BCRYPT_SALT_SIZE)
+            throw new IllegalArgumentException("Salt lenght must be exactly " + BCryptPassword.BCRYPT_SALT_SIZE + " bytes");
```
Here and also above, please use curly braces.

FroMage replied:
If this is required, why doesn't the formatter do that, then?

gsmet commented on extensions/elytron-security-common/runtime/src/main/java/io/quarkus/elytron/security/common/runtime/BcryptUtil.java:
```diff
@@ -0,0 +1,90 @@
+        Objects.requireNonNull(password, "password is required");
+        Objects.requireNonNull(salt, "salt is required");
+        if (salt.length != BCryptPassword.BCRYPT_SALT_SIZE)
+            throw new IllegalArgumentException("Salt lenght must be exactly " + BCryptPassword.BCRYPT_SALT_SIZE + " bytes");
```
```suggestion
            throw new IllegalArgumentException("Salt length must be exactly " + BCryptPassword.BCRYPT_SALT_SIZE + " bytes");
```

gsmet commented on extensions/panache/hibernate-orm-panache/deployment/src/main/java/io/quarkus/hibernate/orm/panache/deployment/PanacheResourceProcessor.java:
```diff
@@ -115,6 +117,8 @@ void build(CombinedIndexBuildItem index,
             transformers.produce(new BytecodeTransformerBuildItem(modelClass, modelEnhancer));
         }
+        if (!modelClasses.isEmpty())
+            entityClasses.produce(new PanacheEntityClassesBuildItem(modelClasses));
```
Same here about curly braces. Maybe configure your Quarkus workspace for it?

gsmet commented on extensions/panache/mongodb-panache/deployment/src/main/java/io/quarkus/mongodb/panache/deployment/PanacheResourceProcessor.java:
```diff
@@ -163,6 +165,8 @@ void build(CombinedIndexBuildItem index,
             propertyMappingClass.produce(new PropertyMappingClassBuildStep(modelClass));
         }
+        if (!modelClasses.isEmpty())
+            entityClasses.produce(new PanacheEntityClassesBuildItem(modelClasses));
```
And here again.

gsmet commented on extensions/pom.xml:
```diff
@@ -97,6 +97,7 @@
 
         <!-- Security -->
         <module>security</module>
+        <module>security-jpa</module>
```
```suggestion
        <module>elytron-security-jpa</module>
```

gsmet commented on extensions/security-jpa/deployment/pom.xml:
```diff
@@ -0,0 +1,114 @@
+    </parent>
+
+    <artifactId>quarkus-security-jpa-deployment</artifactId>
+    <name>Quarkus - Security JPA - Deployment</name>
```
```suggestion
    <name>Quarkus - Elytron Security JPA - Deployment</name>
```
(btw, I agree it's ugly and nobody cares about Elytron but that's how all the others are)

gsmet commented on extensions/security-jpa/deployment/src/main/java/io/quarkus/security/jpa/deployment/JpaSecurityDefinition.java:
```diff
@@ -0,0 +1,125 @@
+        // if it's a panache field, we won't see the getter but it will be there
+        String methodName = "get" + JavaBeanUtil.capitalize(field.name());
+        if (isPanache)
+            return MethodInfo.create(field.declaringClass(), methodName, new Type[0], field.type(), (short) Modifier.PUBLIC);
```
Curly braces.

gsmet commented on extensions/security-jpa/deployment/src/main/java/io/quarkus/security/jpa/deployment/JpaSecurityDefinition.java:
```diff
@@ -0,0 +1,125 @@
+                if (method != null)
+                    return method;
+            }
+        }
```
Curly braces in all this method.

gsmet commented on extensions/security-jpa/deployment/src/main/java/io/quarkus/security/jpa/deployment/QuarkusSecurityJpaProcessor.java:
```diff
@@ -0,0 +1,339 @@
+            BuildProducer<GeneratedBeanBuildItem> beanProducer,
+            List<PanacheEntityClassesBuildItem> panacheEntityClasses) throws Exception {
+
+        // Generate an IdentityProvider if we have a @UserDefinition
```
Same in this class about the curly braces. Please configure your IDE once and for all as it's a recurring issue in your PRs.

FroMage replied:
If it's a requirement the formatter should do it.

gsmet commented on extensions/security-jpa/runtime/pom.xml:
```diff
@@ -0,0 +1,68 @@
+    </parent>
+
+    <artifactId>quarkus-security-jpa</artifactId>
+    <name>Quarkus - Security JPA - Runtime</name>
```
```suggestion
    <name>Quarkus - Elytron Security JPA - Runtime</name>
```

gsmet commented on extensions/security-jpa/runtime/src/main/java/io/quarkus/security/jpa/runtime/Password.java:
```diff
@@ -0,0 +1,26 @@
+ * </p>
+ * <p>
+ * Defaults to considering the password as hashed with bcrypt in the Modular Crypt Format.
+ * </p>
```
For the record, Javadoc complains about the closing `</p>`.

FroMage replied:
I can't imagine why.

dmlloyd replied:
I guess because the end tag is optional.

FroMage replied:
What nonsense.

dmlloyd replied:
It doesn't help matters that IntelliJ aggressively creates end tags...

gsmet commented on extensions/security-jpa/runtime/src/main/resources/META-INF/quarkus-extension.yaml:
```diff
@@ -0,0 +1,12 @@
+---
+name: "Quarkus Security JPA Support"
```
```suggestion
name: "Elytron Security JPA"
```

gsmet reviewed (dismissed):
Some comments are purely cosmetic but some are really important and needs fixing.

FroMage commented:
I've put changes for every comment @gsmet made except the elytron/not debate. I'll let @gsmet and @stuartwdouglas decide here and I'll made any eventual change required.

geoand commented:
At last CI is not throwing random errors at us!

FroMage commented:
FINALLY :) Now we're "just" waiting for @gsmet and @stuartwdouglas to agree on an extension name.

FroMage commented:
@gsmet and @stuartwdouglas any conclusion?

stuartwdouglas commented:
I don't think it should have Elytron in the name, as it does not provide an Elytron realm.

FroMage commented:
In that case we're all good?

gsmet commented on extensions/security-jpa/runtime/src/main/resources/META-INF/quarkus-extension.yaml:
```diff
@@ -0,0 +1,12 @@
+---
+name: "Quarkus Security JPA"
```
```suggestion
name: "Security JPA"
```
Given what I see on `code.quarkus.io`, it shouldn't have Quarkus in the name.

gsmet reviewed (changes requested):
If you can just adjust the extension name and squash everything, let's merge it.

FroMage commented:
Rebased and squashed.

gsmet commented:
Merged, thanks for your patience!

FroMage commented:
Great, thanks!
